speechbrain/nnet/CNN.py

```diff
@@ -1109,6 +1111,440 @@ def forward(self, x):
+
+    def _manage_padding(self, x, kernel_size):
+        # this is the logic that gives correct shape that complies
+        # with the original implementation at https://github.com/google-research/leaf-audio
+
+        def get_padding_value(kernel_size):
+            kernel_sizes = (kernel_size,)
+            from functools import reduce
+            from operator import __add__
+
+            conv_padding = reduce(
+                __add__,
+                [
+                    (k // 2 + (k - 2 * (k // 2)) - 1, k // 2)
+                    for k in kernel_sizes[::-1]
+                ],
+            )
+            return conv_padding
+
+        pad_value = get_padding_value(kernel_size)
+        x = F.pad(x, pad_value, mode=self.padding_mode, value=0)
+        return x
+
+    def _mel_filters(self):
+        def _mel_filters_areas(filters):
+            peaks, _ = torch.max(filters, dim=1, keepdim=True)
+            return (
+                peaks
+                * (torch.sum((filters > 0).float(), dim=1, keepdim=True) + 2)
+                * np.pi
+                / self.n_fft
+            )
+
+        mel_filters = torchaudio.functional.melscale_fbanks(
+            n_freqs=self.n_fft // 2 + 1,
+            f_min=self.min_freq,
+            f_max=self.max_freq,
+            n_mels=self.filters,
+            sample_rate=self.sample_rate,
+        )
+        mel_filters = mel_filters.transpose(1, 0)
+        if self.normalize_energy:
+            mel_filters = mel_filters / _mel_filters_areas(mel_filters)
+        return mel_filters
+
+    def _gabor_params_from_mels(self):
+        coeff = torch.sqrt(2.0 * torch.log(torch.tensor(2.0))) * self.n_fft
+        sqrt_filters = torch.sqrt(self._mel_filters())
+        center_frequencies = torch.argmax(sqrt_filters, dim=1)
+        peaks, _ = torch.max(sqrt_filters, dim=1, keepdim=True)
+        half_magnitudes = peaks / 2.0
+        fwhms = torch.sum((sqrt_filters >= half_magnitudes).float(), dim=1)
+        output = torch.cat(
+            [
+                (center_frequencies * 2 * np.pi / self.n_fft).unsqueeze(1),
+                (coeff / (np.pi * fwhms)).unsqueeze(1),
+            ],
+            dim=-1,
+        )
+        return output
+
+    def _initialize_kernel(self):
+        return self._gabor_params_from_mels()
+
+    def _check_input_shape(self, shape):
+        """Checks the input shape and returns the number of input channels.
+        """
+
+        if len(shape) == 2:
+            in_channels = 1
+        elif len(shape) == 3:
+            in_channels = 1
+        else:
+            raise ValueError(
+                "GaborConv1d expects 2d or 3d inputs. Got " + str(len(shape))
+            )
+
+        # Kernel size must be odd
+        if self.kernel_size % 2 == 0:
+            raise ValueError(
+                "The field kernel size must be an odd number. Got %s."
+                % (self.kernel_size)
+            )
+        return in_channels
+
+
+class Leaf(nn.Module):
```

Comment:
Hi @SarthakYadav, shouldn't this be a lobes instead ? I see it as a "complex" composition rather than a building block.

Reply:
Sure, makes sense. I simply followed SincNet (which was a Module). I'll make it a lobe.